Generate Taylor series

hersle opened the pull request:
Is something like this suited for Symbolics.jl?

hersle commented:
One problem is that the coefficients come out as floats (I want them to be rationals) because of `substitute(..., x => 0)`. A minimal example:
```julia
julia> @variables x
1-element Vector{Num}:
 x

julia> y = expand_derivatives(Differential(x)(sin(x)) / 2)
(1//2)*cos(x)

julia> substitute(y, x => 0)
0.5
```
Using `fold` does not help:
```julia
julia> substitute(y, x => 0, fold=false)
(1//2)*cos(0)

julia> substitute(y, x => 0, fold=true)
0.5
```
Is there an easy way to fix this?

ChrisRackauckas commented on src/taylor.jl:
```diff
@@ -0,0 +1,31 @@
+function taylor(f, x, n::Int)
+    D = Differential(x)
+    n! = factorial(n)
+    c = (D^n)(f) / n!
```
```suggestion
    c = (D^n)(f) // n!
```

?

hersle replied:
Doesn't help with the floats, unfortunately.

ChrisRackauckas commented:
This does seem like a good thing for Symbolics.

@shashi do you know how to make substitute not convert to float?

hersle commented:
One (far from perfect) solution is to `rationalize()` float coefficients into rational numbers. This seems to work well for the tests, except for e.g. the first irrational term in the series for `acos(x) = π/2 + ...`, which becomes `122925461//78256779`.

hersle commented:
There is also [TaylorSeries.jl](https://github.com/JuliaDiff/TaylorSeries.jl), which [could perhaps compose with Symbolics.jl](https://github.com/JuliaDiff/TaylorSeries.jl/issues/340).

hersle commented:
I would appreciate it if someone looks at this :) Suggestions are welcome! Particularly regarding how you think the public Taylor series interface should work, i.e. the current `series`, `taylor` and `taylor_coeff` functions.

karlwessel commented:
I love to take a more closer look at this, but not before mid of next week.

karlwessel commented on src/taylor.jl:
````diff
@@ -0,0 +1,126 @@
+z[0] + z[1]*(-2 + ϵ) + z[2]*((-2 + ϵ)^2) + z[3]*((-2 + ϵ)^3)
+```
+"""
+function series(y, x, x0, ns; name = nameof(y))
````
This won't work for making a taylor series for functions, like for the perturbation theory for differential equations example. Maybe adding a method like this would work:
```julia
series(coeff, x) = sum([x^(i-1)*y for (i, y) in enumerate(coeff)])
```
That could be used like
```julia
@variables t, (y(t))[1:3] epsilon
series(y, epsilon)
```

I think that function could also be used to implement all of the other use cases.
- series with $x_0 \ne 0$: `series(y .+ x0, epsilon)`
- series with first coeff zero: `series([0; y], epsilon)`

hersle replied:
Thanks for this suggestion. I added something similar to let the user both pass their own coefficients, or a variable to automatically define them.

karlwessel reviewed (commented):
I am mixed about the changes. A lot of the explanations in the text have become much clearer, especially on describing perturbation theory and its advantages (for CAS and in general).
It also does a good job to show of the new implemented features for working with Taylor series.

However, since the previous version had to implement all of the taylor specific features by hand it was easier for me to understand the procedure and adapt it for my own problem. Now if one of the new library methods fails for my own specific problem I wouldn't really know what to do. That of course wouldn't be a problem if the new methods just work for everything.

karlwessel commented:
I would suggest to split this PR into multiple steps/stages. 

In the first step introduce and implement easy to understand simple versions of the Taylor series methods in the perturbation example, specific to the perturbation example.

In the next step see how those methods can be used or adapted to work for the example on perturbation theory for ODEs in ModelingToolkit.

And last, if those methods have proven to work in the examples and maybe have been used to solve other perturbation problems (multivariate equations, PDEs, field equations) we can use that experience to decide how and where to add those methods to the Symbolics library or its ecosystem.

hersle commented on docs/src/tutorials/perturbation.md:
````diff
@@ -0,0 +1,141 @@
+```
+According to Abel's theorem, a general quintic equation does not have a closed form solution. But we can easily solve it numerically using Newton's method (here implemented for simplicity, and not performance):
+```@example perturb
+function solve_newton(eq, x, x₀; abstol=1e-8, maxiters=50)
````
I agree with this from a pedagogical viewpoint. I must have gotten ahead of myself when I realized the satisfying general procedure 😅

hersle replied:
I removed/eased up the automatic perturbation solver in the most recent tutorial, so it is more similar to the original.

hersle commented:
Thanks for the feedback, I agree with many of your points. Maybe splitting the changes according to your 3 incremental stages would be the best approach.

> What does Symbolics + this PR provide that wouldn't be easily doable with TaylorSeries?

Ideally, I agree it would be fantastic to compose with TaylorSeries.jl. That should work for series with numerical coefficients (e.g. `exp(x) = 1.0 + 1.0*x + ...`), but maybe not as well for series with symbolic coefficients, as the tutorial examples? This is the most important concern for Symbolics, in my opinion.

> since the previous version had to implement all of the taylor specific features by hand it was easier for me to understand the procedure and adapt it for my own problem.

I can relate to this. But I think if Symbolics is to be a viable CAS, it is more important to have the expected convenience functions so every user does not have to reimplement them. It is also on the wanted feature list (#59). 

I must leave this as-is for now. It works and is "complete", but can surely be improved. If I get time, maybe I will get back at it. Anyone is welcome to take over, of course.

ChrisRackauckas commented:
The additional functionality is definitely a yes: yes people can build this functionality from pure Julia but of course that's the case because the library is all Julia. That's no reason to not improve the user experience. That doesn't mean we can't have tutorials of people doing this stuff by hand, but most people shouldn't need to.

Let's get this rebased and tests passing. I think the only thing left is to potentially write out a few more steps in the perturbation tutorial.

hersle commented:
Alright I will try to do it in the near future.

hersle commented:
I rebased and incorporated some of @karlwessel's suggestions. Hopefully this reaches a balance point.

I'm not 100% convinced the current `series()`, `taylor()`, `taylor_coeff()`, `...` interface is optimal, but hopefully it is a decent starting point. I don't feel strongly about it, as long as it is simple, consistent and convenient, so anyone is welcome to change it.

karlwessel commented:
Thanks for finalizing this! I wasn't aware this was on the wanted features list.
